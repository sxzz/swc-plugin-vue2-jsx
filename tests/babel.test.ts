import { describe, expect, test } from 'vitest'
import { transformJsx } from './_util'

const tests = [
  {
    name: 'HTML tag',
    from: `render(h => <div>test</div>)`,
    to: `render(h => h("div", ["test"]));`,
  },
  {
    name: 'HTML tag if variable in scope',
    from: `const div = {}; render(h => <div>test</div>)`,
    to: `const div = {};
render(h => h("div", ["test"]));`,
  },
  {
    name: 'Tag & Component',
    from: `const Alpha = {}; render(h => [<Alpha>test</Alpha>, <Beta>test</Beta>])`,
    to: `const Alpha = {};
render(h => [h(Alpha, ["test"]), h("Beta", ["test"])]);`,
  },
  {
    name: 'MemberExpression component',
    from: `render(h => <a.b.c>test</a.b.c>)`,
    to: `render(h => h(a.b.c, ["test"]));`,
  },
  {
    name: 'Variable as content',
    from: `render(h => <div>{test}</div>)`,
    to: `render(h => h("div", [test]));`,
  },
  {
    name: 'Spread content',
    from: `render(h => <div>{...test}</div>)`,
    to: `render(h => h("div", [...test]));`,
  },
  {
    name: 'Self-closing html tag',
    from: `render(h => <div><br/></div>)`,
    to: `render(h => h("div", [h("br")]));`,
  },
  {
    name: 'Combined content',
    from: `render(h => <div>
  test{test} {...test}
  <tag1 />
  <tag2 />

  Some text
  goes here


</div>)`,
    to: `render(h => h("div", ["test", test, " ", ...test, h("tag1"), h("tag2"), "Some text goes here"]));`,
  },
  {
    name: 'Plain attrs',
    from: `render(h => <div id="hi" dir="ltr"></div>)`,
    to: `render(h => h("div", {
  "attrs": {
    "id": "hi",
    "dir": "ltr"
  }
}));`,
  },
  {
    name: 'Expression attrs',
    from: `render(h => <div id={id}></div>)`,
    to: `render(h => h("div", {
  "attrs": {
    "id": id
  }
}));`,
  },
  {
    name: 'Special root attributes',
    from: `render(h => (
      <div
        class="foo"
        style="bar"
        key="key"
        ref="ref"
        refInFor
        slot="slot"
        model={{
          value: this.txt,
          callback: $$v => {
            this.txt = $$v
          }
        }}>
      </div>
    ))`,
    to: `render(h => h("div", {
  "class": "foo",
  "style": "bar",
  "key": "key",
  "ref": "ref",
  "refInFor": true,
  "slot": "slot",
  "model": {
    value: this.txt,
    callback: $$v => {
      this.txt = $$v;
    }
  }
}));`,
  },
  {
    name: 'Special snake-case attributes',
    from: `render(h => (
      <div
        props-on-success={noop}
        on-click={noop}
        on-kebab-case={noop}
        domProps-innerHTML="<p>hi</p>"
        hook-insert={noop}>
      </div>
    ))`,
    to: `render(h => h("div", {
  "props": {
    "on-success": noop
  },
  "on": {
    "click": noop,
    "kebab-case": noop
  },
  "domProps": {
    "innerHTML": "<p>hi</p>"
  },
  "hook": {
    "insert": noop
  }
}));`,
  },
  {
    name: 'Special camelCase attributes',
    from: `render(h => (
      <div
        propsOnSuccess={noop}
        onClick={noop}
        onCamelCase={noop}
        domPropsInnerHTML="<p>hi</p>"
        hookInsert={noop}>
      </div>
    ))`,
    to: `render(h => h("div", {
  "props": {
    "onSuccess": noop
  },
  "on": {
    "click": noop,
    "camelCase": noop
  },
  "domProps": {
    "innerHTML": "<p>hi</p>"
  },
  "hook": {
    "insert": noop
  }
}));`,
  },
  {
    name: 'Data attributes',
    from: `render(h => (
      <div data-id="1"></div>
    ))`,
    to: `render(h => h("div", {
  "attrs": {
    "data-id": "1"
  }
}));`,
  },
  {
    name: 'Inline spread attributes',
    from: `const props = {
      innerHTML: 2
    }
    const otherName = {
      someAttr: 'hey'
    }
    const vnode = render(h => (
      <div propsHello="test" {...{ props, attrs: otherName }}/>
    ))`,
    to: `const props = {
  innerHTML: 2
};
const otherName = {
  someAttr: 'hey'
};
const vnode = render(h => h("div", {
  "props": {
    "hello": "test",
    ...props
  },
  "attrs": { ...otherName
  }
}));`,
  },
  {
    name: 'Only spread attributes',
    from: `render(h => <div {...spread} />)`,
    to: `import _mergeJSXProps from "@vue/babel-helper-vue-jsx-merge-props";
render(h => h("div", _mergeJSXProps([{}, spread])));`,
  },
  {
    name: 'Generic spread attributes',
    from: `render(h => <div {...spread} hello="world" />)`,
    to: `import _mergeJSXProps from "@vue/babel-helper-vue-jsx-merge-props";
render(h => h("div", _mergeJSXProps([{}, spread, {
  "attrs": {
    "hello": "world"
  }
}])));`,
  },
  //   {
  //     name: 'Directives',
  //     from: `render(h => <div v-test={ 123 } vSomething_modifier={ 1234 } vOtherStuff:argument_modifier1_modifier2={ 234 } />)`,
  //     to: `render(h => h("div", {
  //   "directives": [{
  //     name: "test",
  //     value: 123
  //   }, {
  //     name: "something",
  //     value: 1234,
  //     modifiers: {
  //       "modifier": true
  //     }
  //   }, {
  //     name: "other-stuff",
  //     value: 234,
  //     arg: "argument",
  //     modifiers: {
  //       "modifier1": true,
  //       "modifier2": true
  //     }
  //   }]
  // }));`,
  //   },
  {
    name: 'xlink',
    from: `<use xlinkHref={'#name'} />`,
    to: `h("use", {
  "attrs": {
    "xlink:href": '#name'
  }
});`,
  },
  {
    name: 'Magic domProps input[value]',
    from: `<input type="text" value={val} />`,
    to: `h("input", {
  "attrs": {
    "type": "text"
  },
  "domProps": {
    "value": val
  }
});`,
  },
  {
    name: 'Magic domProps option[selected]',
    from: `<option selected={val} />`,
    to: `h("option", {
  "domProps": {
    "selected": val
  }
});`,
  },
  {
    name: 'Magic domProps input[checked]',
    from: `<input checked={val} />`,
    to: `h("input", {
  "domProps": {
    "checked": val
  }
});`,
  },
  {
    name: 'Magic domProps video[muted]',
    from: `<video muted={val} />`,
    to: `h("video", {
  "domProps": {
    "muted": val
  }
});`,
  },
  {
    name: 'Multiple event listeners',
    from: `<div onClick={listner1} onClick={listner2} onClick={listner3} />`,
    to: `h("div", {
  "on": {
    "click": [listner1, listner2, listner3]
  }
});`,
  },
  {
    name: 'Root attribute',
    from: `<MyComponent propsProp1="foo" props={{ prop1: 'alpha', prop2: 'beta' }} />`,
    to: `import _mergeJSXProps from "@vue/babel-helper-vue-jsx-merge-props";
h("MyComponent", _mergeJSXProps([{
  "props": {
    "prop1": "foo"
  }
}, {
  "props": {
    prop1: 'alpha',
    prop2: 'beta'
  }
}]));`,
  },
  {
    name: 'JSX comments',
    from: `<div><p>jsx</p>{/* <p>comment</p> */}</div>`,
    to: `h("div", [h("p", ["jsx"])]);`,
  },
  {
    name: 'Underscore Props',
    from: `const MyComp = {}; render(h => <MyComponent attrs-my_prop="test">test</MyComponent>)`,
    to: `const MyComp = {};
render(h => h("MyComponent", {
  "attrs": {
    "my_prop": "test"
  }
}, ["test"]));`,
  },
]

describe('from babel jsx', async () => {
  for (const { name, from } of tests) {
    test(name, async () => {
      expect(await transformJsx(from)).toMatchSnapshot()
    })
  }
})
