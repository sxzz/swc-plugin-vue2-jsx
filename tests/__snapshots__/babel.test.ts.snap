// Vitest Snapshot v1

exports[`from babel jsx > Combined content 1`] = `
"render((h)=>h(\\"div\\", [
        \\"test\\",
        test,
        \\" \\",
        ...test,
        h(\\"tag1\\"),
        h(\\"tag2\\"),
        \\"Some text goes here\\"
    ]));
"
`;

exports[`from babel jsx > Data attributes 1`] = `
"render((h)=>h(\\"div\\", {
        attrs: {
            \\"data-id\\": \\"1\\"
        }
    }));
"
`;

exports[`from babel jsx > Expression attrs 1`] = `
"render((h)=>h(\\"div\\", {
        attrs: {
            \\"id\\": id
        }
    }));
"
`;

exports[`from babel jsx > Generic spread attributes 1`] = `
"import _mergeJSXProps from \\"@vue/babel-helper-vue-jsx-merge-props\\";
render((h)=>h(\\"div\\", _mergeJSXProps([
        spread,
        {
            attrs: {
                \\"hello\\": \\"world\\"
            }
        }
    ])));
"
`;

exports[`from babel jsx > HTML tag 1`] = `
"render((h)=>h(\\"div\\", [
        \\"test\\"
    ]));
"
`;

exports[`from babel jsx > HTML tag if variable in scope 1`] = `
"const div = {};
render((h)=>h(\\"div\\", [
        \\"test\\"
    ]));
"
`;

exports[`from babel jsx > Inline spread attributes 1`] = `
"import _mergeJSXProps from \\"@vue/babel-helper-vue-jsx-merge-props\\";
const props = {
    innerHTML: 2
};
const otherName = {
    someAttr: 'hey'
};
const vnode = render((h)=>h(\\"div\\", _mergeJSXProps([
        {
            props: {
                \\"hello\\": \\"test\\"
            }
        },
        {
            props,
            attrs: otherName
        }
    ])));
"
`;

exports[`from babel jsx > JSX comments 1`] = `
"h(\\"div\\", [
    h(\\"p\\", [
        \\"jsx\\"
    ])
]);
"
`;

exports[`from babel jsx > Magic domProps input[checked] 1`] = `
"h(\\"input\\", {
    domProps: {
        \\"checked\\": val
    }
});
"
`;

exports[`from babel jsx > Magic domProps input[value] 1`] = `
"h(\\"input\\", {
    attrs: {
        \\"type\\": \\"text\\"
    },
    domProps: {
        \\"value\\": val
    }
});
"
`;

exports[`from babel jsx > Magic domProps option[selected] 1`] = `
"h(\\"option\\", {
    domProps: {
        \\"selected\\": val
    }
});
"
`;

exports[`from babel jsx > Magic domProps video[muted] 1`] = `
"h(\\"video\\", {
    domProps: {
        \\"muted\\": val
    }
});
"
`;

exports[`from babel jsx > MemberExpression component 1`] = `
"render((h)=>h(a.b.c, [
        \\"test\\"
    ]));
"
`;

exports[`from babel jsx > Multiple event listeners 1`] = `
"h(\\"div\\", {
    on: {
        \\"click\\": [
            listner1,
            listner2,
            listner3
        ]
    }
});
"
`;

exports[`from babel jsx > Only spread attributes 1`] = `
"render((h)=>h(\\"div\\", spread));
"
`;

exports[`from babel jsx > Plain attrs 1`] = `
"render((h)=>h(\\"div\\", {
        attrs: {
            \\"id\\": \\"hi\\",
            \\"dir\\": \\"ltr\\"
        }
    }));
"
`;

exports[`from babel jsx > Root attribute 1`] = `
"h(\\"MyComponent\\", {
    props: {
        \\"prop1\\": \\"foo\\",
        \\"\\": {
            prop1: 'alpha',
            prop2: 'beta'
        }
    }
});
"
`;

exports[`from babel jsx > Self-closing html tag 1`] = `
"render((h)=>h(\\"div\\", [
        h(\\"br\\")
    ]));
"
`;

exports[`from babel jsx > Special camelCase attributes 1`] = `
"render((h)=>h(\\"div\\", {
        props: {
            \\"onSuccess\\": noop
        },
        on: {
            \\"click\\": noop,
            \\"camelCase\\": noop
        },
        domProps: {
            \\"innerHTML\\": \\"<p>hi</p>\\"
        },
        hook: {
            \\"insert\\": noop
        }
    }));
"
`;

exports[`from babel jsx > Special root attributes 1`] = `
"render((h)=>h(\\"div\\", {
        class: \\"foo\\",
        style: \\"bar\\",
        key: \\"key\\",
        ref: \\"ref\\",
        refInFor: true,
        slot: \\"slot\\",
        model: {
            value: this.txt,
            callback: ($$v)=>{
                this.txt = $$v;
            }
        }
    }));
"
`;

exports[`from babel jsx > Special snake-case attributes 1`] = `
"render((h)=>h(\\"div\\", {
        props: {
            \\"on-success\\": noop
        },
        on: {
            \\"click\\": noop,
            \\"kebab-case\\": noop
        },
        domProps: {
            \\"innerHTML\\": \\"<p>hi</p>\\"
        },
        hook: {
            \\"insert\\": noop
        }
    }));
"
`;

exports[`from babel jsx > Spread content 1`] = `
"render((h)=>h(\\"div\\", [
        ...test
    ]));
"
`;

exports[`from babel jsx > Tag & Component 1`] = `
"const Alpha = {};
render((h)=>[
        h(\\"Alpha\\", [
            \\"test\\"
        ]),
        h(\\"Beta\\", [
            \\"test\\"
        ])
    ]);
"
`;

exports[`from babel jsx > Underscore Props 1`] = `
"const MyComp = {};
render((h)=>h(\\"MyComponent\\", {
        attrs: {
            \\"my_prop\\": \\"test\\"
        }
    }, [
        \\"test\\"
    ]));
"
`;

exports[`from babel jsx > Variable as content 1`] = `
"render((h)=>h(\\"div\\", [
        test
    ]));
"
`;

exports[`from babel jsx > xlink 1`] = `
"h(\\"use\\", {
    attrs: {
        \\"xlink:href\\": '#name'
    }
});
"
`;
